Add tests

lausobo opened the pull request:
Backpressure works well until there is a broken window or a timeout (last two test cases)

lausobo committed:
Add some basic tests

lausobo committed:
Add backpressure tests, two failiing

akarnokd commented:
I've updated the code of the operator:

```java
package com.luissolano;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import io.reactivex.internal.util.BackpressureHelper;
import org.reactivestreams.*;

import io.reactivex.*;
import io.reactivex.Scheduler.Worker;
import io.reactivex.disposables.*;
import io.reactivex.schedulers.Schedulers;

/* Imported from http://stackoverflow.com/questions/42114546/rxjava-substitute-a-subsequence-of-elements-with-a-single-element/42117531#42117531
    author David Karnok (https://github.com/akarnokd)
 */
public class Main {

    public static void main(String[] args) {
        Flowable<String> source = Flowable.just(
                "A", "A", "R", "S", "A", "R", "F", "R", "A", "A");

        source.lift(new ConditionalCompactor(
                500, TimeUnit.SECONDS, Schedulers.computation()))
                .subscribe(System.out::println, Throwable::printStackTrace);

    }

    static final class ConditionalCompactor implements FlowableOperator<String, String> {
        final Scheduler scheduler;

        final long timeout;

        final TimeUnit unit;

        ConditionalCompactor(long timeout, TimeUnit unit,
                             Scheduler scheduler) {
            this.scheduler = scheduler;
            this.timeout = timeout;
            this.unit = unit;
        }

        @Override
        public Subscriber<? super String> apply(Subscriber<? super String> t) {
            return new ConditionalCompactorSubscriber(
                    t, timeout, unit, scheduler.createWorker());
        }

        static final class ConditionalCompactorSubscriber
                implements Subscriber<String>, Subscription {
            final Subscriber<? super String> actual;

            final Worker worker;

            final long timeout;

            final TimeUnit unit;

            final AtomicInteger wip;

            final SerialDisposable mas;

            final Queue<String> queue;

            final List<String> batch;

            final AtomicLong requested;

            Subscription s;

            static final Disposable NO_TIMER;
            static {
                NO_TIMER = Disposables.empty();
                NO_TIMER.dispose();
            }

            volatile boolean done;
            Throwable error;

            boolean compacting;

            int lastLength;

            ConditionalCompactorSubscriber(Subscriber<? super String> actual,
                                           long timeout, TimeUnit unit, Worker worker) {
                this.actual = actual;
                this.worker = worker;
                this.timeout = timeout;
                this.unit = unit;
                this.batch = new ArrayList<>();
                this.wip = new AtomicInteger();
                this.mas = new SerialDisposable();
                this.mas.set(NO_TIMER);
                this.queue = new ConcurrentLinkedQueue<>();
                this.requested = new AtomicLong();
            }

            @Override
            public void onSubscribe(Subscription s) {
                this.s = s;
                actual.onSubscribe(this);
            }

            @Override
            public void onNext(String t) {
                queue.offer(t);
                drain();
            }

            @Override
            public void onError(Throwable e) {
                error = e;
                done = true;
                drain();
            }

            @Override
            public void onComplete() {
                done = true;
                drain();
            }

            @Override
            public void cancel() {
                s.cancel();
                worker.dispose();
            }

            @Override
            public void request(long n) {
                BackpressureHelper.add(requested, n);
                s.request(n);
                drain();
            }

            void drain() {
                if (wip.getAndIncrement() != 0) {
                    return;
                }
                int missed = 1;
                for (;;) {

                    long r = requested.get();
                    long e = 0L;

                    while (e != r) {
                        boolean d = done;
                        if (d && error != null) {
                            queue.clear();
                            actual.onError(error);
                            worker.dispose();
                            return;
                        }
                        String s = queue.peek();
                        if (s == null) {
                            if (d) {
                                actual.onComplete();
                                worker.dispose();
                                return;
                            }
                            break;
                        }

                        if (compacting) {
                            batch.clear();
                            batch.addAll(queue);
                            int n = batch.size();
                            String last = batch.get(n - 1);
                            if ("S".equals(last)) {
                                if (n > 1) {
                                    actual.onNext(queue.poll());
                                    mas.set(NO_TIMER);
                                    lastLength = -1;
                                    compacting = false;
                                    e++;
                                    continue;
                                }
                                // keep the last as the start of the new
                                if (lastLength <= 0) {
                                    lastLength = 1;
                                    mas.set(worker.schedule(() -> {
                                        queue.offer("T");
                                        drain();
                                    }, timeout, unit));
                                    this.s.request(1);
                                }
                                break;
                            } else
                            if ("T".equals(last)) {
                                actual.onNext(queue.poll());
                                compacting = false;
                                mas.set(NO_TIMER);
                                lastLength = -1;
                                e++;
                                continue;
                            } else
                            if ("F".equals(last)) {
                                actual.onNext("M");
                                while (n-- != 0) {
                                    queue.poll();
                                }
                                compacting = false;
                                mas.set(NO_TIMER);
                                lastLength = -1;
                                e++;
                            } else {
                                if (lastLength != n) {
                                    lastLength = n;
                                    mas.set(worker.schedule(() -> {
                                        queue.offer("T");
                                        drain();
                                    }, timeout, unit));
                                    this.s.request(1);
                                }
                                break;
                            }
                        } else {
                            if ("A".equals(s) || "F".equals(s) || "R".equals(s)) {
                                queue.poll();
                                actual.onNext(s);
                                e++;
                            } else
                            if ("T".equals(s)) {
                                // ignore timeout markers outside the compacting mode
                                queue.poll();
                            } else {
                                compacting = true;
                            }
                        }
                    }

                    if (e != 0L) {
                        BackpressureHelper.produced(requested, e);
                    }

                    if (e == r) {
                        if (done) {
                            if (error != null) {
                                queue.clear();
                                actual.onError(error);
                                worker.dispose();
                                return;
                            }
                            if (queue.isEmpty()) {
                                actual.onComplete();
                                worker.dispose();
                                return;
                            }
                        }
                    }

                    missed = wip.addAndGet(-missed);
                    if (missed == 0) {
                        break;
                    }
                }
            }
        }
    }
}
```

and one of your unit tests needs adjusting because you have to await the response to a delayed emission of a subsequent item:

```java

@Test
public void testItFlushesOnTimeoutAndBackpressure() throws InterruptedException {

    TestSubscriber<String> subscriber = new TestSubscriber<>(0);

    Flowable<String> source = Flowable.concat(
            Flowable.just("A", "A", "R", "S", "A"),
            Flowable.just("R").delay(1, TimeUnit.SECONDS),
            Flowable.just("F", "R")
    );

    source.lift(new Main.ConditionalCompactor(100, TimeUnit.MILLISECONDS, Schedulers.computation()))
            .subscribe(subscriber);

    subscriber.request(1);

    subscriber.assertValues("A");

    subscriber.request(1);

    subscriber.assertValues("A", "A");

    subscriber.request(1);

    subscriber.assertValues("A", "A", "R");

    subscriber.request(1);

    Thread.sleep(600);

    subscriber.assertValues("A", "A", "R", "S");

    Thread.sleep(600);

    subscriber.request(1);

    subscriber.assertValues("A", "A", "R", "S", "A");

    subscriber.request(1);

    subscriber.assertValues("A", "A", "R", "S", "A", "R");

    subscriber.request(1);

    subscriber.assertValues("A", "A", "R", "S", "A", "R", "F");

    subscriber.request(1);

    subscriber.assertValues("A", "A", "R", "S", "A", "R", "F", "R");

    subscriber.assertComplete();
}
```

lausobo committed:
Adjust unit test as per Davids comment

Second part of this comment: https://github.com/luisobo/rxjava-conditional-compacted-operator/pull/2#issuecomment-278732974

luisobo commented:
@akarnokd Pushed the updates to the operator to master and push the update to the test here. All tests pass!
Gonna merge this PR. 

I've got another one coming that generalizes the operator (that much I can do 😅 )

Also, do you want to fork this repo to keep it in your GitHub? This code is all yours 😅 

Thanks again for all your help!
